File: services/geminiService.ts
import { GoogleGenAI } from "@google/genai";

const getAIClient = () => {
  const apiKey = process.env.API_KEY;
  if (!apiKey) {
    console.warn("API Key not found in environment");
    return null;
  }
  return new GoogleGenAI({ apiKey });
};

export const analyzePlantHealth = async (base64Image: string): Promise<string> => {
  const ai = getAIClient();
  if (!ai) return "API Key missing. Cannot analyze.";

  try {
    const response = await ai.models.generateContent({
      model: 'gemini-2.5-flash',
      contents: {
        parts: [
          {
            inlineData: {
              mimeType: 'image/jpeg',
              data: base64Image
            }
          },
          {
            text: "You are an agricultural expert AI. Analyze this plant leaf image. Detect any signs of disease, nutrient deficiency, or water stress. If it looks healthy, say so. Keep the response concise (max 3 sentences)."
          }
        ]
      }
    });
    return response.text || "Analysis failed.";
  } catch (error) {
    console.error("Gemini analysis error:", error);
    return "Error analyzing image. Please try again.";
  }
};

export const analyzeSecurityImage = async (base64Image: string): Promise<string> => {
  const ai = getAIClient();
  if (!ai) return "API Key missing.";

  try {
    const response = await ai.models.generateContent({
      model: 'gemini-2.5-flash',
      contents: {
        parts: [
          {
            inlineData: {
              mimeType: 'image/jpeg',
              data: base64Image
            }
          },
          {
            text: "You are a farm security AI. Identify what caused the motion trigger in this image. Is it a human, an animal, or a false alarm? Be brief."
          }
        ]
      }
    });
    return response.text || "Analysis failed.";
  } catch (error) {
    console.error("Gemini analysis error:", error);
    return "Error analyzing security image.";
  }
};

export const getFarmingAdvice = async (context: string, question: string): Promise<string> => {
    const ai = getAIClient();
    if (!ai) return "API Key missing.";
  
    try {
      const response = await ai.models.generateContent({
        model: 'gemini-2.5-flash',
        contents: `Context: ${context}\n\nUser Question: ${question}\n\nAnswer as a helpful farming assistant:`
      });
      return response.text || "I couldn't process that request.";
    } catch (error) {
      console.error("Gemini chat error:", error);
      return "Error connecting to AI advisor.";
    }
  };

File: Backend/app/services/gemini_service.py
import google.generativeai as genai
from ..config import settings
import base64
from io import BytesIO
from PIL import Image

# Configure Gemini API
genai.configure(api_key=settings.gemini_api_key)


class GeminiService:
    def __init__(self):
        self.model = genai.GenerativeModel('gemini-2.0-flash-exp')
    
    async def analyze_plant_health(self, image_base64: str) -> str:
        """
        Analyze plant leaf image for health issues.
        
        Args:
            image_base64: Base64 encoded image string
            
        Returns:
            Analysis result as text
        """
        try:
            # Decode base64 image
            image_data = base64.b64decode(image_base64)
            image = Image.open(BytesIO(image_data))
            
            prompt = """You are an agricultural expert AI. Analyze this plant leaf image. 
            Detect any signs of disease, nutrient deficiency, or water stress. 
            If it looks healthy, say so. Keep the response concise (max 3 sentences)."""
            
            response = self.model.generate_content([prompt, image])
            return response.text
        
        except Exception as e:
            return f"Error analyzing image: {str(e)}"
    
    async def analyze_security_image(self, image_base64: str) -> str:
        """
        Analyze security camera image for motion detection.
        
        Args:
            image_base64: Base64 encoded image string
            
        Returns:
            Analysis result as text
        """
        try:
            # Decode base64 image
            image_data = base64.b64decode(image_base64)
            image = Image.open(BytesIO(image_data))
            
            prompt = """You are a farm security AI. Identify what caused the motion trigger in this image. 
            Is it a human, an animal, or a false alarm? Be brief."""
            
            response = self.model.generate_content([prompt, image])
            return response.text
        
        except Exception as e:
            return f"Error analyzing security image: {str(e)}"
    
    async def get_farming_advice(self, context: str, question: str) -> str:
        """
        Get farming advice based on context and user question.
        
        Args:
            context: Current farm conditions/context
            question: User's question
            
        Returns:
            AI-generated advice
        """
        try:
            prompt = f"""Context: {context}

User Question: {question}

Answer as a helpful farming assistant:"""
            
            response = self.model.generate_content(prompt)
            return response.text
        
        except Exception as e:
            return f"Error getting advice: {str(e)}"


# Create singleton instance
gemini_service = GeminiService()
